### PR TITLE
Improved leaderboard submission Vegard

### PR DESCRIPTION
**Pull request by vskogstad**
Made some changes. Barely within 90 mins, so could be over on a different H100. 5 out of 90 minutes are spent validating against entire validation set.

**Improvements**
-Changed the data-loader from random, to randomized strided sampling without replacement. More unique training samples gave increased training loss(no repetitions) but brought the validation loss down to 3.126. 
-Implemented gated attention (Sort of like in qwen next, but I do full gates instead of per head and use SILU instead of sigmoid). Changed to just 6 attention heads, which is a bit worse but gives higher MFU. In sum this brought validation loss down to 3.1035. https://arxiv.org/pdf/2505.06708
-QK-norm -> 3.094
-Adjusting AdamW params from [0.90, 0.95] to [0.9, 0.999] gave a surprisingly large boost -> 3.0839
-U-net architecture with learnable params. -> 3.0762

**Other attempts**
-Lowering the LR for the LM-head layer as recommended in Dion. NanoGPT seems to be doing it the other way around? Gave very minimal improvements. https://arxiv.org/pdf/2504.05295
-Document masking. Lower MFU and very slight performance decrease. I really expected this to work a lot better.
-Sliding window attention. Hybrid with 3 layers sliding window 1 full layer is only slightly worse, but no speedup
-Grouped query attention. Worse, as expected.
-Scaling output of each block like in Ernie. Worse. Think this might be interfering with my layernorm scaling.
-Mixing in extra embedding values in later value matrices like in NanoGPT speedrun. (Looks good 2/3 in but ends up worse in the end). General idea: https://arxiv.org/pdf/2410.17897
-Decreasing/increasing warmup steps or increasing learning rate after adding QK-norm gave no benefit.
-QK-clip. I was not able to get this to work. In theory it should help a bit with the MFU compared to QK-norm.

### CODE REVIEW

**Comment by vskogstad** [comment]
Not much progress lately. I've tested implementing model growth and WSD/WSM. 

For WSM I'm able to match WSD, but WSD works so much worse for me than a cosine scheduler. With WSM I am almost running out of storage space on vast.ai with 20+ checkpoints, so probably not going to submit this. 
Model growth is roughly equal to my previous best models when training for 25% of the tokens and using a growth rate of 2. Adds complexity to the pipeline though, might be that with better tuned LR this could be better. Will probably try this if I train a bigger model.

I believe both of these could be good/great for a longer training runs but I am not seeing speedups in 1.5 hour.

**Comment by vskogstad** [comment]
Been working through assignment 2.
Ported over my triton kernel for flash attention to work within my pytorch model. Increased the batch size to 512, using gradient accumulation.
Final validation loss 3.0158 in ~82 mins. 
Closing this and adding a new pull request based on latest version of the list.
